### PR TITLE
add vault_gcp_auth_backend_role data source

### PR DESCRIPTION
**Pull request by davidmontoyago**
When creating a `vault_identity_entity_alias` a `role_id` is required to bind the identity to a role. When using `vault_gcp_auth_backend_role` there is no way to fetch the `role_id` which leaves us unable to use Identities with GCP backends. This PR adds the data source for `vault_gcp_auth_backend_role` with acc tests and docs.

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

Relates OR Closes #0000

Release note for [CHANGELOG](https://github.com/hashicorp/terraform-provider-vault/blob/master/CHANGELOG.md):

```release-note
new data source `vault_gcp_auth_backend_role`
```

Output from acceptance testing:

```
TF_ACC=1 go test -v github.com/hashicorp/terraform-provider-vault/vault -run TestAccGCPAuthBackendRoleDataSource_basic

=== RUN   TestAccGCPAuthBackendRoleDataSource_basic
2021/03/23 17:28:28 [INFO] terraform: building graph: GraphTypeValidate
2021/03/23 17:28:28 [INFO] terraform: building graph: GraphTypeRefresh
....

DESTROY: data.vault_gcp_auth_backend_role.gcp_role

STATE:

data.vault_gcp_auth_backend_role.gcp_role:
  ID =
  provider = provider.vault
2021/03/23 17:28:42 [INFO] terraform: building graph: GraphTypeApply
2021/03/23 17:28:42 DestroyEdgeTransformer: pruning unused resource node data.vault_gcp_auth_backend_role.gcp_role (prepare state)
2021/03/23 17:28:44 [INFO] terraform: building graph: GraphTypePlanDestroy
2021/03/23 17:28:44 [INFO] terraform: building graph: GraphTypeRefresh
2021/03/23 17:28:45 [INFO] terraform: building graph: GraphTypePlanDestroy
--- PASS: TestAccGCPAuthBackendRoleDataSource_basic (16.94s)
PASS
ok      github.com/hashicorp/terraform-provider-vault/vault     17.561s
...
```

### CODE REVIEW

**Comment by davidmontoyago** [comment]
@catsby @tvoran hi friends, sorry for the ping. Any chance I can get a review? Our team is currently unable to manage Vault Identities with GCP auth backends.

**Review by jasonodonnell — CHANGES_REQUESTED** [review]
Hi @davidmontoyago, thanks for this contribution!

In general this looks good, however, there's a bug causing a panic with `bound_labels`:

```
panic: interface conversion: interface {} is map[string]interface {}, not *schema.Set

goroutine 84 [running]:
github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*MapFieldWriter).setSet(0xc000719be0, 0xc0004e5350, 0x1, 0x1, 0x1df0e40, 0xc00070e030, 0xc0005dd900, 0x1f91003, 0xc000128b40)
	/Users/jasonodonnell/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.9.0/helper/schema/field_writer_map.go:344 +0xc05
github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*MapFieldWriter).set(0xc000719be0, 0xc0004e5350, 0x1, 0x1, 0x1df0e40, 0xc00070e030, 0x1, 0x1)
	/Users/jasonodonnell/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.9.0/helper/schema/field_writer_map.go:107 +0x250
github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*MapFieldWriter).WriteField(0xc000719be0, 0xc0004e5350, 0x1, 0x1, 0x1df0e40, 0xc00070e030, 0x0, 0x0)
	/Users/jasonodonnell/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.9.0/helper/schema/field_writer_map.go:89 +0x3f9
github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*ResourceData).Set(0xc000158070, 0x1f91003, 0xc, 0x1df0e40, 0xc00070e030, 0x1, 0x0)
	/Users/jasonodonnell/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.9.0/helper/schema/resource_data.go:199 +0x1cf
github.com/hashicorp/terraform-provider-vault/vault.gcpAuthBackendRoleRead(0xc000158070, 0x1f6c8a0, 0xc0006ee0e0, 0xc000158070, 0x0)
	/Users/jasonodonnell/Git/terraform-provider-vault/vault/data_source_gcp_auth_backend_role.go:132 +0x485
github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*Resource).ReadDataApply(0xc0004bbdd0, 0xc000719040, 0x1f6c8a0, 0xc0006ee0e0, 0xc000184680, 0x1, 0x0)
	/Users/jasonodonnell/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.9.0/helper/schema/resource.go:403 +0x88
github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*Provider).ReadDataApply(0xc000290080, 0xc00031fad0, 0xc000719040, 0xc000719040, 0x0, 0x0)
	/Users/jasonodonnell/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.9.0/helper/schema/provider.go:451 +0x8f
github.com/hashicorp/terraform-plugin-sdk/internal/helper/plugin.(*GRPCProviderServer).ReadDataSource(0xc000288230, 0x217aea0, 0xc0005d63f0, 0xc00035e100, 0xc000288230, 0xc0005d63f0, 0xc000a31b78)
	/Users/jasonodonnell/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.9.0/internal/helper/plugin/grpc_provider.go:1036 +0x42d
github.com/hashicorp/terraform-plugin-sdk/internal/tfplugin5._Provider_ReadDataSource_Handler(0x1f46f80, 0xc000288230, 0x217aea0, 0xc0005d63f0, 0xc0007b60c0, 0x0, 0x217aea0, 0xc0005d63f0, 0xc00026e000, 0x15c)
	/Users/jasonodonnell/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.9.0/internal/tfplugin5/tfplugin5.pb.go:3341 +0x214
google.golang.org/grpc.(*Server).processUnaryRPC(0xc000102900, 0x2186bc0, 0xc000295680, 0xc0006fc100, 0xc0005d6a20, 0x28bbe50, 0x0, 0x0, 0x0)
	/Users/jasonodonnell/go/pkg/mod/google.golang.org/grpc@v1.27.1/server.go:1024 +0x522
google.golang.org/grpc.(*Server).handleStream(0xc000102900, 0x2186bc0, 0xc000295680, 0xc0006fc100, 0x0)
	/Users/jasonodonnell/go/pkg/mod/google.golang.org/grpc@v1.27.1/server.go:1313 +0xd34
google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc000602180, 0xc000102900, 0x2186bc0, 0xc000295680, 0xc0006fc100)
	/Users/jasonodonnell/go/pkg/mod/google.golang.org/grpc@v1.27.1/server.go:722 +0xa5
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/Users/jasonodonnell/go/pkg/mod/google.golang.org/grpc@v1.27.1/server.go:720 +0xa5

Error: The terraform-provider-vault plugin crashed!
```

Bound labels is a little weird because you send a list to Vault, but it returns a map. This case will need to be handled specifically to avoid a panic when its set.

**Comment by davidmontoyago** [comment]
@jasonodonnell ack! thank you so much for reviewing. Will reproduce and patch today.

File: vault/data_source_gcp_auth_backend_role.go [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+	for _, k := range gcpRoleFields {
+		if v, ok := resp.Data[k]; ok {
+			if err := d.Set(k, v); err != nil {
+				return fmt.Errorf("error reading %s for GCP Auth Backend Role %q: %q", k, path, err)
+			}
+		}
+	}
+
+	// bound_labels is configured as list but returned as map
+	if v, ok := resp.Data["bound_labels"]; ok {
```
Comment (davidmontoyago) [inline vault/data_source_gcp_auth_backend_role.go:138]: This is just copy/paste from [esource_gcp_auth_backend_role.go#L359](https://github.com/hashicorp/terraform-provider-vault/blob/master/vault/resource_gcp_auth_backend_role.go#L359). Should we DRY it up into a `readBoundLabelsField` func? I'm not sure it's worth it

**Comment by davidmontoyago** [comment]
@jasonodonnell test case and patch are in.

```
TF_ACC=1 go test -v github.com/hashicorp/terraform-provider-vault/vault -run TestAccGCPAuthBackendRoleDataSource_gce

=== RUN   TestAccGCPAuthBackendRoleDataSource_gce
2021/04/19 17:31:47 DestroyEdgeTransformer: pruning unused resource node data.vault_gcp_auth_backend_role.gcp_role (prepare state)
2021/04/19 17:32:24 DestroyEdgeTransformer: pruning unused resource node data.vault_gcp_auth_backend_role.gcp_role (prepare state)
--- PASS: TestAccGCPAuthBackendRoleDataSource_gce (99.39s)
PASS
ok      github.com/hashicorp/terraform-provider-vault/vault     100.045s
```

```
TF_ACC=1 go test -v github.com/hashicorp/terraform-provider-vault/vault -run TestAccGCPAuthBackendRoleDataSource_basic

=== RUN   TestAccGCPAuthBackendRoleDataSource_basic
2021/04/19 17:33:57 DestroyEdgeTransformer: pruning unused resource node data.vault_gcp_auth_backend_role.gcp_role (prepare state)
2021/04/19 17:34:10 DestroyEdgeTransformer: pruning unused resource node data.vault_gcp_auth_backend_role.gcp_role (prepare state)
--- PASS: TestAccGCPAuthBackendRoleDataSource_basic (28.72s)
PASS
ok      github.com/hashicorp/terraform-provider-vault/vault     29.884s
```

**Comment by herrewig** [comment]
@jasonodonnell is it possible to take another look at this PR since @davidmontoyago has implemented your requested change? Thank you for everything - we sincerely appreciate your time.

**Comment by davidmontoyago** [comment]
@jasonodonnell sorry for the ping. we went to production with this a while back. Please lend us a review when possible. Thank you and your team for the awesome Vault tooling.

**Review by jasonodonnell — APPROVED** [review]
LGTM, thank you for your patience and the contribution!

**Pull request merged**